scripts/terminalserver/trees.jl

```diff
@@ -0,0 +1,160 @@
+
+function treerender(x)
+    fields = fieldnames(typeof(x))
+
+    if isempty(fields)
+        treerender(Text(string(typeof(x), "()")))
+    else
+        treerender(LazyTree(string(typeof(x)), function ()
+            [SubTree(string(f), getfield_safe(x, f)) for f in fields]
+        end))
+    end
+end
+
+function treerender(x::AbstractDict)
+    treerender(LazyTree(string(nameof(typeof(x)), "{$(eltype(x).parameters[1]), $(eltype(x).parameters[2])}"), function ()
+        if length(keys(x)) > 25
+            partition_by_keys(x)
+        else
+            [SubTree(repr(k), x[k]) for k in keys(x)]
+        end
+    end))
+end
+
+function treerender(x::Module)
+    treerender(LazyTree(string(x), function ()
+        ns = names(x, all = true)
+        out = []
+        for n in ns
+            isdefined(x, n) || continue
+            Base.isdeprecated(x, n) && continue
+            startswith(string(n), '#') && continue
+
+            v = getfield(x, n)
+            v === x && continue
+
+            push!(out, SubTree(string(n), v))
+        end
+
+        out
+    end))
+end
+
+function treerender(x::AbstractArray)
+    treerender(LazyTree(string(typeof(x)), function ()
+        if length(x) > 50
+            partition_by_keys(x)
+        else
+            vec(x)
+        end
+    end))
+end
+
+treerender(x::Number) = treerender(Leaf(strlimit(repr(x), limit=100)))
+treerender(x::AbstractString) = treerender(Leaf(strlimit(repr(x), limit=100)))
+treerender(x::AbstractChar) = treerender(Leaf(strlimit(repr(x), limit=100)))
+treerender(x::Symbol) = treerender(Leaf(strlimit(repr(x), limit=100)))
+treerender(x::Nothing) = treerender(Leaf(strlimit(repr(x), limit=100)))
+treerender(x::Missing) = treerender(Leaf(strlimit(repr(x), limit=100)))
+treerender(x::Ptr) = treerender(Leaf(string(typeof(x), ": 0x", string(UInt(x), base=16, pad=Sys.WORD_SIZE>>2))))
+treerender(x::Text) = treerender(Leaf(x.content))
+treerender(x::Function) = treerender(Leaf(string(x)))
+
+function partition_by_keys(x, _keys = keys(x); sz = 20, maxparts = 100)
+    partitions = Iterators.partition(_keys, max(sz, length(_keys) ÷ maxparts))
+    out = []
+    for part in partitions
+        if length(part) > sz
+            push!(out, LazyTree(string(first(part), " ... ", last(part)), function ()
+                partition_by_keys(x, part, sz = sz, maxparts = maxparts)
+            end))
+        else
+            push!(out, LazyTree(string(first(part), " ... ", last(part)), function ()
+                [SubTree(repr(k), x[k]) for k in part]
+            end))
+        end
+    end
+    return out
+end
+
+function get_lazy(id::Int)
+    try
+        if haskey(TREES, id)
+            x = [treerender(x) for x in pop!(TREES, id).children()]
+            return x
+        else
+            return ["[out of date result]"]
+        end
+    catch err
+        @error exception=(err, catch_backtrace())
+        return ["nope"]
+    end
```

Comment:
Can this case ever happen ? Well, I think we should keep this anyway though.

Reply:
I don't think so, but let's see if someone complains about it ;)

Reply:
good 👍